dev-docs/RFCs/v7.x/gpu-grid-layer.md

```diff
@@ -0,0 +1,40 @@
+# RFC: GPU Acceleration in GridLayer
+
+* **Authors**: Ravi Akkenapally
+* **Date**: May 2019
+* **Status**: **Draft**
+
+
+## Abstract
+
+GPU accelerated aggregation is currently available in `ScreenGridLayer` and `ContourLayer`. In 6.3 release, `GPUAggregator` is enhanced to support aggregation of multiple weights required for `GridLayer`. This RFC proposes adding GPU accelerated aggregation to `GridLayer`.
+
+
+## Fallback cases
+
+There are several features of `GridLayer` that can't be supported by GPU Aggregation.
+
+* `Lower and higher Percentile` : When these props are set to a non zero value, grid cells are sorted based on their aggregated values and some of the begining and ending cells are ignored (not rendered) based on specified percentile value. On GPU, sorting is not possible.
+
+* `Aggregation operation` : Any aggregation other than `Min`, `Max`, `SUM` or `MEAN` is not supported on GPU.
+
+* `Picking` : When aggregation is performed on CPU, each cell can point to an array of objects that fall into the cell. And when a cell is picked, `GridLayer` can provide this array of objects. But on GPU picking information will be limited to `aggregated value` of the cell and `domain value` (min/max value of all aggregated cells).
+
+* `Custom Color/Elevation calculations` : `GridLayer` has accessor props `getColorValue` and `getElevationValue`, which take an array of all points fall into a grid cell and calculate the `color` and `elevation` of that cell. Not all custom implementation of these accessors will be supported, but most common usecases will be supported. More details below.
+
+* `WebGL` : GPU Aggregaiton is only supported when using `WebGL2` context, when using `WebGL` context we fallback to CPU Aggregation.
+
+
+## Color and Elevation Accessors
+
+`GridLayer`'s `getColor` and `getElevation` perform two things. 1. Specify which field in original data set to use as weight and 2. aggregation operation. To support GPU Aggregation, these accessor props need to be split into two new props.
+
+    1.`getColorWeight/getElevationWeight`: specifies weight field of each data point.
```

Comment:
This is just a possible idea (perhaps for the longer term): Rather than having each aggregating layer define a (growing) list of specific aggregations it may be worth considering a more general aggregator that lets the user specify a list of aggregations they want to do.

```
  const aggregator = new Aggregator({
    aggregations: {
      COUNT_SUM: {aggregation: 'SUM', field: 'count'},
      VALUE_MEAN: {aggregation: 'MEAN', field: 'value'}
    }
  });
```    

Then the layer could just refer to the aggregations

```
   color: aggregator.COUNT_SUM,
   colorScale: ..., // Some format that can be GLSL compiled
   elevation: aggregator.VALUE_MIN,
   ...
```

This is just pseudo-code as I am not very familiar with the actual details. But again, my general sense would be to generalize and share as much as possible between the aggregating layers.

Reply:
This is a good idea, to provide a consistent API. But each layer has fixed number of aggregation,   for example, `GridLayer` has two (color and elevation), `ScreenGridLayer` has one and `ContourLayer` has one. I feel, instead of grouping them into single class, we can have just one prop for each aggregation, for example , in `GridLayer` :
```
colorAggregation: {operation: 'SUM', weight: 'count'} 
elevationAggregation: {operation: 'MAX', weight: x => x.SPACES}
```
and default value for above props will be `{operation: 'SUM', weight: 1}`

Hear `operation` defines how to aggregate values, and `weight` can be `Number`, `String` (used as key to access from the data point) or `Function`, takes original data object as input and returns weight.

May be there is a value defining a single class as above, I will add it in the RFC and we can discuss for any future use.

Reply:
One use case for separating out the aggregations is that it would now be possible to reuse the same aggregation for multiple attributes (which should be faster than doing multiple aggregations?).

Perhaps we could also feed multiple layers from the same aggregations, instead of having to aggregate separately for each layer?

Ideally, the aggregation class eould be completely independent of the layer, so that any layer could be driven by the output of the aggregation class, instead of some layers being carefully hard coded to support aggregation.

This may not be achievable, but I think it could be a useful perspective to have during the API design.

Reply:
Just an idea:

```elevationAggregation: {operation: 'MAX', weight: x => x.SPACES}```

It may be useful to support string weights 

```elevationAggregation: {operation: 'MAX', weight: `SPACES`}```

This would be similar to how `@deck.gl/json` supports accessors.

The idea being that if this is not a JavaScript function we could do something with it in GLSL (an alternative would be to "scrape" the javascript function definition).

Reply:
Yep, as mentioned above, `weight` can be a `Number`, `String` or a `Function`.